### ISSUE TITLE
Propose to taint node "shutdown" condition 

### ISSUE THREAD

**Issue by jingxu97** [kubernetes/kubernetes#58635]
Today there is only one node condition to represent whether kubelet is healthy (updating node status). Node controller will mark node as "NotReady" after kubelet stops updating for a while. The reason for kubelet not responding could be 
1. Network disconnect
2. Node is shun down (stopped) or paused
3. Kubelet is killed (could not response)

## Problem
In many cases, when a node becomes "NotReady", pods will be evicted, and new pods will be started on other nodes (failover for high availability). However, if node is shut down, the volumes attached to the nodes are still attached. Volume controller should try to detach volumes when pods are killed from the old node. But there is a safety check before detaching to see whether volume is still mounted. This check is done by a VolumeInUse field in node status which is updated by kubelet. In case of node shut down, this field will not be updated by kubelet so that the safety check will fail and assume the volume cannot be detached because of existing mounts. The new pods will be pending for a while because volume cannot be attached to the other node. Although the reality is that volume can be detached in this situation. (Note: some cloud provider will treat stopped node as not exist in cloud provider so that node controller will delete the node API object. But still detach will not happen because of the mount safety check. Ref https://github.com/kubernetes/kubernetes/issues/46442)

## Proposal
We propose to taint the node with "Stopped" condition. When kubelet stops updating, node controller checks cloud provider and see whether the node is in "stopped" state (Note: different cloud provider might use different term to represent this state, e.g., terminated). When volume controller gets a node update/delete event, it can check the node taint spec and mark the volumes attached to that node are safe to detach (volumes can be be removed from the list of VolumeInUse (meaning the volumes are mounted). When detach is issued, the operation can be proceeded without delay.

To determine the "stopped" condition, I lists a few cloud provider's explanation of stopping an instance.

### GCE

- Stopping an instance causes Compute Engine to send the ACPI Power Off signal to the instance. Modern guest operating systems are configured to perform a clean shutdown before powering off in response to the power off signal. Compute Engine waits a short time for the guest to finish shutting down and then transitions the instance to the TERMINATED state.

- You can stop an instance temporarily so you can come back to it at a later time. A stopped instance does not incur charges, but all of the resources that are attached to the instance will still be charged. Alternatively, if you are done using an instance, delete the instance and its resources to stop incurring charges.

- You can permanently delete an instance to remove the instance and the associated resources from your project. If the instance is part of an instance group, the group might try to recreate the instance to maintain a certain group size.

### AWS

- You can only stop an Amazon EBS-backed instance. The instance performs a normal shutdown and stops running; its status changes to stopping and then stopped. Any Amazon EBS volumes remain attached to the instance, and their data persists. In most cases, the instance is migrated to a new underlying host computer when it's started. 

- You can delete your instance when you no longer need it. This is referred to as terminating your instance. When an instance terminates, the data on any instance store volumes associated with that instance is deleted. If your instance is in an Auto Scaling group, the Auto Scaling service marks the stopped instance as unhealthy, and may terminate it and launch a replacement instance.

### OpenStack

- Admins can pause and unpause a Nova compute instance. When an instance is paused, the entire state of the instance is kept in RAM. Pausing an instance will disable access to that instance, but won't free up any of its resources. Another option is to suspend, and then resume, an instance. Like paused OpenStack instances, a suspended instance keeps its current state, but it is written to storage. 

- A third option is to shelve OpenStack instances.A shelved instance is actually shut down, which is not the case for suspended or paused instances. If admins decide they no longer need a shelved instance, they can remove it, which ensures that it doesn't maintain any hypervisor-level resources in use.

- The last option is to stop an instance in Nova, which will disconnect all of its associated resources. This means admins can't restore a stopped instance to its previous state. This option is only useful for OpenStack instances that an organization no longer needs. In all other cases, admins should shelve, suspend or pause the instance.

So basically a "stopped" node means that it is shut down like an operating system shut down situation. All mounts are gone when machine is shut down.

cc @yujuhong @dchen1107 @Random-Liu

**Comment by k82cn** [comment]
xref https://github.com/kubernetes/kubernetes/issues/49798

**Comment by yujuhong** [comment]
I talked to @bsalamat and @dchen1107 and I think it's advised to use taints instead of conditions (ref: #57690).

That said, I find the definition of "stopped" a bit vague. @jingxu97, could you define what "stopped" means, whether a node can recover from the state, and if it can, what do we expect the state of the recovered node to be. Also, it'd be nice to see how "stopped" would be interpreted in different cloud providers while adhering to the definition.

**Comment by zetaab** [comment]
@yujuhong stopped means for instance virtual machine in openstack which is stopped (not running currently). Currently this thing is not supported at all, and if we stop instance that node is deleted from kubernetes but volumes are still there as attached.

**Comment by jingxu97** [comment]
@yujuhong @bsalamat and @dchen1107 I updated the proposal based on the feedback. PTAL Thanks!

**Comment by yastij** [comment]
@jingxu97 @yujuhong  +1 the way to go is taint, as work is being done to move our internal logic to taints instead of condition.

Also Looping @augabet @EtienneCoutaud

**Comment by bsalamat** [comment]
There are some ambiguities here for me. To clarify things and see if my understanding is correct, let's look at different scenarios one at a time (I use node and the underlying VM of the node interchangeably here):
1. Node has network issues and cannot be reached: This is recoverable and I don't think "stopped" would be the right taint for it. It is after all not stopped. It is not responding to network requests. Our control plane can check with the cloud provider API whether the underlying VM of the node is still running.
1. Kubelet has died or not responding. This is more or less similar to the above case. I don't think "stopped" is the right taint.
1. Node is paused. I think "stopped" (or "paused") taint would be appropriate here.
1. Node is shutdown. IIUC, when a node's underlying VM is terminated, the node is deleted from Kubernetes. So, there is nothing to taint and we must detach the volumes. Am I missing something?

So, among the four cases, one of them looks to me like a candidate for "paused" taint.

**Comment by jingxu97** [comment]
@bsalamat Thanks for your comments. For scenarios 3 and 4, I think there is some confusion.
3. Node is paused (sleep/suspend): actually in this case, since OS is not shutdown, the mounts will still exist, so for our case, we cannot detach volumes, which means we should not taint the node with situation
4. Node is shutdown: when node is shutdown, actually the node is still exist in cloud provider (e.g., GCE, AWS), their status is terminated/stopped. But see issue https://github.com/kubernetes/kubernetes/issues/46442, the cloud provider code in kubernetes handles this situation differently. Some provider (GCE) will not delete the node from Kuberentes, some others will.  We want to taint the node even if they will be deleted, so that volume controller can check the node taint and know whether mounts are still exist or not. When node is deleted, detach will be triggered, but the problem is that we don't know whether mounts still exist or not and will not issue detach until some timeout. If node is taint with "shutdown/stopped", we can be sure. It is ok that node will be deleted as long as it is tainted before.
(Note, for GCE, if node is managed by instance group, the shutdown node will be actually deleted from cloud provider and a new one comes up to replace it) 

Please let me know if you have questions about it. Thanks!

**Comment by bsalamat** [comment]
@jingxu97 Thanks for clarification. I guess I now understand it better. So, the taint will be applied only to scenario #4 where the node is shutdown, because the node is not deleted from K8s clusters in all cloud providers. This sounds good to me. Maybe, we want to call the taint "shutdown" then?

**Comment by yujuhong** [comment]
I've already seen many different operations like "suspend", "pause", "stop", "shelve"...
Regardless what the final name of the taint is, I think it's fair to say that the taint should clearly state that the node should be in the shutdown state (e.g., volumes unmounted), etc.

**Comment by yastij** [comment]
@jingxu97 @bsalamat @yujuhong - before moving with this, we should clarify how we should handle quick reboot use cases, where users rely on the fact volumes are still attached.

One solution is to add a flag to handle this and not break expected behaviour.

**Comment by jingxu97** [comment]
@yastij For quick reboot case, the following will happen, I think.
1. After node is shutdown for about half a minute, node controller will check cloud provider whether node still exist or not. If node at this moment, is already restarted, node controller will not do anything (except updating node status)
2. If node is not yet up when node controller checks cloud provider, currently, for some cloud provider (AWS etc) node object will be deleted from API server when node is not in running state. Pods might be soon garbage collected because node no long exists (Again, if node comes back soon, Pod garbage collector might not delete pods). When pods are deleted, volume controller will try to detach volume but it will fail because it assumes mounts are still exist. For some other cloud provider (GCE etc), node object will not be deleted if node still exist in cloud provider (the state is changed to terminated or something). Then after  that, depending on the taint and eviction policy, pods might be evicted after sometime. When pods are evicted, volume detach operation again will fail first because controller cannot determine mounts exist or not.

Now with node taint "shutdown" state, the above scenario will be the same, if node is rebooted very quickly (less than 30 seconds or so), nothing will happen. If node is rebooted after a couple of minutes, depending on the cloud provider and also the timing of node reboot, pods might be garbage collected/evicted. With node taint, detach operation can be triggered because controller can know the mounts are gone and safe to detach. 

So my point is "taint" will not change the current behavior of pods deletion. It can help volume controller to determine whether to detach volume safely.

**Comment by yastij** [comment]
@jingxu97 - ok I'm fine with, so basically what we have to do:

1. Create the shutdown taint
2. Let the cloud/node_controller taint the nodes that have been terminated
3. Make volume controller look for shutdown taint and unmount volumes

Will this be transparent to CSI or should it watch for the taint and issue corresponding call to the plugin ?

**Comment by jingxu97** [comment]
This will be transparent to CSI. Only volume controller needs some changes to check the taint so that detach will be issued without delay.

**Comment by yastij** [comment]
@jingxu97 - SGTM

**Comment by jberkus** [comment]
@jingxu97 if this change is going to be in 1.10, please add kind/ and priority/ labels to it.  Also, this looks remarkably like a feature which wasn't submitted for feature freeze.  Please correct, thanks!

**Comment by jingxu97** [comment]
We are trying to find some help implementing this. Can I still submit a feature for this? Thanks!

**Comment by yastij** [comment]
I can help with this

**Comment by jingxu97** [comment]
@yastij that's great! Please let me know if you need anything. As @jberkus mentioned, I think we might need to submit a feature.

**Comment by yastij** [comment]
@jingxu97 - I can submit it tomorrow and start working on a PR, is it targeting 1.10 ?

**Comment by jingxu97** [comment]
Yes, we want to target 1.10.

**Comment by jhorwit2** [comment]
cc @cheftako @wlan0

**Comment by davidopp** [comment]
@gmarek

**Comment by liggitt** [comment]
Just to be clear, what component is intended to add and remove the taint?

**Comment by zetaab** [comment]
@liggitt I think cloud controller should add that taint? Instead of deleting node here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L238-L239 we add taint if node is in shutdown state. The problem currently is that we might need extend Instances() in cloudproviders. I am thinking to add InstanceShutdownedByProviderID  here https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/cloud.go#L149

so instead of going to delete node here https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L254 we query InstanceShutdownedByProviderID() and if instance is in shutdown state lets add taint and continue to next item in that loop.

But I do not know where we should delete that taint if node comes back? Possible in that same function?

**Comment by liggitt** [comment]
> @liggitt I think cloud controller should add that taint?

Ok, just wanted to make sure the kubelet wasn't expected to taint/untaint itself as part of this

**Comment by yastij** [comment]
cc @wlan0  @cheftako are you okay with extending instances interface ? This will break some of the CCM already implemented, are we okay with that ?

@zetaab - on the same func would be fine, as it is used to monitor nodes

**Comment by zetaab** [comment]
I coded something regarding this, could you please comment is this approach the correct one and can it be accepted? If yes, then I can make PR of it :) https://github.com/zetaab/kubernetes/commit/1cca2e424b61976b0e6086332c4e108b1b8ae73c

Example using this cloud-controller-manager in openstack:
```
% kubectl get nodes
NAME                  STATUS    ROLES     AGE       VERSION
kube-dev-master-1-0   Ready     <none>    2m        v1.9.1
kube-dev-node-1-0     Ready     <none>    39s       v1.9.1
kube-dev-node-1-1     Ready     <none>    39s       v1.9.1
% kubectl describe node kube-dev-node-1-0
...
Taints:             <none>

then shutdown node using openstack API, after ~30seconds:

% kubectl get nodes
NAME                  STATUS     ROLES     AGE       VERSION
kube-dev-master-1-0   Ready      <none>    3m        v1.9.1
kube-dev-node-1-0     NotReady   <none>    2m        v1.9.1
kube-dev-node-1-1     Ready      <none>    2m        v1.9.1
% kubectl describe node kube-dev-node-1-0
...
Taints:             node.cloudprovider.kubernetes.io/shutdowned=true:NoSchedule

start instance again from openstack API, after 30-60seconds:

% kubectl get nodes
NAME                  STATUS    ROLES     AGE       VERSION
kube-dev-master-1-0   Ready     <none>    5m        v1.9.1
kube-dev-node-1-0     Ready     <none>    3m        v1.9.1
kube-dev-node-1-1     Ready     <none>    3m        v1.9.1
% kubectl describe node kube-dev-node-1-0
...
Taints:             <none>

```

Questions/comments:
- can we modify Instances interface?
- I need help implementing at least gce/aws part of InstanceShutdownedByProviderID
- the main point why we need this is because of volume detach things. Should that be separate PR or also in this one? After this node label thing is added we should somehow query node taints https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L202 here. If node has shutdown taint, we can skip waiting time. How we should query that? Query node state from apiserver using kubeclient? @jingxu97 
- this needs cloud-controller-manager in use. However, the current situation products like kops / openshift those are not using cloud-controller-manager. If cloud-controller-manager is not in use, the taints are not added.

**Comment by zetaab** [comment]
I opened PR lets continue there, #59323

**Comment by gmarek** [comment]
+1 this sounds exactly like a thing Taints are for. I'd go with a name more connected to the reasoning behind the Taint (sth. like DisksUnmounted or sth like that), than a generic "Node is shut down" case. Unless someone can give me a guarantee that it's the only case we'd like to force detachment. Together with the Taint we'll need to have an admission plugin that will add a toleration for this taint to Pods that don't use persistent volumes.

As for more design-ish ideas I think there's no question about what component should apply the Taint (side note - it should be NoExecute, not NoSchedule one).

I'm not that convinced about taint removal though. Initially it needs to be NodeController, because version skew (i.e. Kubelets can lag up to 2 versions behind the control plane), so that's certainly a initial requirement. This leads to a question when we want to do that:
1. when Kubelet starts reporting 'Ready' again
1. when NC notices that the VM is not in 'terminated' state anymore

If the answer is 1 then this goes back to the discussion about Kubelet applying Taints itself, that I don't have capacity to hold right now, but I believe we'll need to get back to it sooner or later.

**Comment by khenidak** [comment]
I am against taints in cloud controller. The cloud controller should provide/extend existing instance interface with *IsRunning(nodeName)*. This can be translated by the node controller to taints/actions as needed. If the cloud provider put the taint then it needs to maintain a reference to tainted nodes and update them accordingly (i.e node came back up, node is not coming back up etc).

Keep in mind that a node (from cloud perspective) might be part of the user provisioned nodes but no longer part of the cluster.

**Comment by jingxu97** [comment]
@gmarek Currently node "shut down" seems the only situation. After os shut down, the mounts are gone so detach can proceed. Otherwise, we will rely kubelet to update node status to indicate whether mounts still exist or not. In case of node deleted from cloud provider, volumes are automatically detached. 

I also agree with @khenidak. I think it is better to extend cloud provider interface to have IsShutdown(nodeName) and let the node lifecycle controller takes care of the taint.

**Comment by jingxu97** [comment]
reopen the issue since the PR is revered and opened again.

**Comment by dims** [comment]
since we are not going to make progress in v1.10, i'll clear the milestone. please respond if this is wrong.

/milestone clear

**Comment by yastij** [comment]
cc @jingxu97 @gmarek

**Comment by gmarek** [comment]
I believe it's already too late for that.

**Comment by cheftako** [comment]
@mtaufen and I noticed that the actual cloud provider InstanceShutdownByProviderID() method is not yet implemented by any in tree cloud provider. As such it seems like the taint will never be set for any in tree provider. Is this issue really resolved?

**Comment by yastij** [comment]
@cheftako - currently none of the in-tree provider implements it. Also notes that the volume controller do not currently implements the logic of dettaching volumes when nodeShutdown taint is applied.

**Comment by yastij** [comment]
Yes this targets 1.12

**Comment by cheftako** [comment]
Part of this is adding cloud-provider specific code to pkg/controller/nodelifecycle/node_lifecycle_controller.go. We are trying to remove all such cloud provider dependencies from core K8s/K8s code in 1.12. Can we work out a way to do this without changing node_lifecycle_controller? Making the changes in pkg/controller/cloud or equivalent CCM controller seems preferable.

**Comment by k82cn** [comment]
Really a long history :)

> Yes this targets 1.12

@yastij , what's scope/subtasks for 1.12, e.g. supported cloud providers? alpha or GA?

> Can we work out a way to do this without changing node_lifecycle_controller? Making the changes in pkg/controller/cloud or equivalent CCM controller seems preferable.

As @cheftako mentioned, are we on the same page for the involved components? Taint nodes by `node_life_cycle` vs. `CCM/cloud`.

**Comment by adampl** [comment]
Does anyone know whether it can also work with bare-metal nodes? I mean, if a node were tainted manually (or by some external software) with `nodeShutdown`, would the volumes be automatically detached?
